base/linalg/dense.jl

```diff
@@ -358,6 +358,10 @@ kron(a::AbstractVector, b::AbstractMatrix) = kron(reshape(a, length(a), 1), b)
 
 # Matrix power
 (^)(A::AbstractMatrix, p::Integer) = p < 0 ? Base.power_by_squaring(inv(A), -p) : Base.power_by_squaring(A, p)
+function (^)(A::AbstractMatrix{T}, p::Integer) where T<:Integer
+    TT = Base.promote_op(^, T, typeof(p))
```

Comment:
I guess the previous discussion got hidden by github: https://github.com/JuliaLang/julia/pull/23368#discussion_r134434571
The conversion is needed to make sure that Integer promotion rules carry over to matrices. For example, without this conversion `[1 1;1 0]^big(10000)` would return `Matrix{Int64}` (and negative values) rather than `Matrix{BigInt}`. This would be inconsistent with the behavior of `2^big(10000)`, which returns `BigInt`.